python/mujoco_mpc/demos/direct/direct_optimizer.py

```diff
@@ -0,0 +1,1168 @@
+        if model.sensor_needstage[i] == mujoco.mjtStage.mjSTAGE_POS:
+          adr = model.sensor_adr[i]
+          dim = model.sensor_dim[i]
+          idx = slice(adr, adr + dim)
+          sensor[idx, t] = data.sensordata[idx]
+
+    # last step
+    elif t == horizon - 1:
+      # set data
+      data.qpos = qpos[:, t]
+      data.qvel = qvel[:, t]
+      data.qacc = np.zeros(model.nv)
+
+      # evaluate position and velocity sensors
+      mujoco.mj_fwdPosition(model, data)
+      mujoco.mj_sensorPos(model, data)
+      mujoco.mj_fwdVelocity(model, data)
+      mujoco.mj_sensorVel(model, data)
+
+      # zero memory
+      sensor[:, t] = 0.0
+
+      # only set position and velocity sensors
+      for i in range(model.nsensor):
+        needstage = model.sensor_needstage[i]
+        if (
+            needstage == mujoco.mjtStage.mjSTAGE_POS
+            or needstage == mujoco.mjtStage.mjSTAGE_VEL
+        ):
+          adr = model.sensor_adr[i]
+          dim = model.sensor_dim[i]
+          idx = slice(adr, adr + dim)
+          sensor[idx, t] = data.sensordata[idx]
+    else:
+      # set data
+      data.qpos = qpos[:, t]
+      data.qvel = qvel[:, t]
+      data.qacc = qacc[:, t]
+
+      # inverse dynamics
+      mujoco.mj_inverse(model, data)
+
+      # copy sensor and force
+      sensor[:, t] = data.sensordata
+      force[:, t] = data.qfrc_inverse
+
+  return sensor, force
+
+
+# %%
+# inverse dynamics (derivatives wrt qpos, qvel, qacc)
+# dfdq, dfdv, dfda, dsdq, dsdv, dsda
+def diff_inverse_dynamics(
+    model: mujoco.MjModel,
+    data: mujoco.MjData,
+    qpos: npt.ArrayLike,
+    qvel: npt.ArrayLike,
+    qacc: npt.ArrayLike,
+    horizon: int,
+    eps: float = 1.0e-8,
+    flg_actuation: bool = True,
+) -> Tuple[
+    npt.ArrayLike,
+    npt.ArrayLike,
+    npt.ArrayLike,
+    npt.ArrayLike,
+    npt.ArrayLike,
+    npt.ArrayLike,
+]:
+  # Jacobians
+  dfdq = [np.zeros((model.nv, model.nv)) for _ in range(horizon)]
+  dfdv = [np.zeros((model.nv, model.nv)) for _ in range(horizon)]
+  dfda = [np.zeros((model.nv, model.nv)) for _ in range(horizon)]
+  dsdq = [np.zeros((model.nsensordata, model.nv)) for _ in range(horizon)]
+  dsdv = [np.zeros((model.nsensordata, model.nv)) for _ in range(horizon)]
+  dsda = [np.zeros((model.nsensordata, model.nv)) for _ in range(horizon)]
+
+  # transposed Jacobians
+  dqds = np.zeros((model.nv, model.nsensordata))
+  dvds = np.zeros((model.nv, model.nsensordata))
+  dads = np.zeros((model.nv, model.nsensordata))
+  dqdf = np.zeros((model.nv, model.nv))
+  dvdf = np.zeros((model.nv, model.nv))
+  dadf = np.zeros((model.nv, model.nv))
+
+  # loop over horizon
+  for t in range(0, horizon):
+    # first step
+    if t == 0:
+      # set data
+      data.qpos = qpos[:, t]
+      data.qvel = np.zeros(model.nv)
+      data.qacc = np.zeros(model.nv)
+
+      # Jacobian
+      mujoco.mjd_inverseFD(
+          model,
+          data,
+          eps,
+          flg_actuation,
+          None,
```

Comment:
I think all these None arguments are optional in the Python bindings so you can delete them and only specify DsDq=dqds.

Same in the other call below.

Reply:
Done.